packages/react-reconciler/src/ReactFiber.js

```diff
@@ -388,10 +388,14 @@ export function createFiberFromElement(
   }
 
   let fiber;
-  const type = element.type;
+  let type = element.type;
   const key = element.key;
   let pendingProps = element.props;
 
+  if (type !== null && type !== undefined && typeof type.then === 'function') {
+    type = resolveThenableType(type);
```

Comment:
I think there's a subtle issue with this approach. Since this throws when the parent reconciles, there is no way to continue siblings of the component that is lazy even if they're available. By making it throw in the beginWork of the individual component, we avoid that problem.

Reply:
I have a test for that. Works by creating a new component type and throwing in beginWork, as you say. Not the greatest, but in the new approach we discussed offline will be better.